[FEAT] MyPage의 선호/비선호 조회, 추가, 삭제 구현

HyeonChae1104 opened the pull request:
## 주요 변경사항 ✔️
### hooks
`useGetFood.ts` : 카테고리별 메뉴를 받아오는 로직
`useGetPreferenceFoods.ts` : 선호음식 조회
`useGetNonPreferenceFoods.ts` : 비선호 음식 조회
`useAddFoodPreference.ts` : 선호/비선호 음식 추가를 위한 훅
`useDeleteFoodPreference.ts` : 음식 삭제를 위한 훅
### components/feature/MyPage
`Preference` : 선호하는 음식
`NonPreference` : 꺼려하는 음식
`FoodSelector` : 
- 사용자가 카테고리별로 음식을 선택할 수 있도록 `CategoryDropdownMenu` 컴포넌트 구현
- 선택한 항목을 적용할 수 있는 `ConfirmButton` 추가
- `useGetCategory`와 `useGetFoodsByCategory` 훅을 사용하여 동적 데이터 가져오기 기능 구현 

## 리뷰어에게... 🙏
며칠동안 계속 안쉬고 했는데도 이제 마무리가 되었네요..
늦어져서 죄송합니다..ㅜㅜ

검색 기능은 저번 PR에는 추가를 해두었는데, 이번 PR에는 아직 들어가지 않았습니다.! 오늘 안에 기능을 추가하도록 하겠습니다!

디자인 부분도 부족할텐데 피드백 부탁드립니다! 바로바로 변경하겠습니다!
## 관련 이슈

closes #95

HyeonChae1104 committed:
feat(hooks): 선호/비선호 음식 조회 구현

- 사용자 선호 음식을 조회하는 `useGetPreferenceFoods` 훅 추가
- 사용자 비선호 음식을 조회하는 `useGetNonPreferenceFoods` 훅 추가

#95

HyeonChae1104 committed:
feat(hooks): 카테고리별 음식 조회 훅 구현

- 카테고리에 따른 음식 조회 `useGetFoodsByCategory` 훅 추가
- `getFoodsByCategory` 함수에서 잘못된 카테고리 예외 처리 추가

#95

HyeonChae1104 committed:
feat(hooks): 음식 추가 훅 구현

- 선호/비선호 음식을 추가할 때 사용하는 훅을 구현하였습니다.

#95

HyeonChae1104 committed:
feat(hooks): 선호/비선호 음식을 삭제하는 훅 구현

-  추가된 선호/비선호 음식을 삭제하는 로직을 훅으로 구현하였습니다.

#95

HyeonChae1104 committed:
feat(mock): 목데이터 구현

- food.mock.ts 는 카테고리, 카테고리별 메뉴, 선호음식 조회, 비선호 음식 조회를 위해 만들어진 목데이터입니다.
- 음식 추가, 삭제 구현을 확인하기 위해 handler.ts를 추가하였습니다.

#95

HyeonChae1104 committed:
feat(type): food type 추가

#95

HyeonChae1104 committed:
feat(Preference): 마이페이지의 PreferenceSection 구현

- 선호하는 음식 섹션을 구현하였습니다.

#95

HyeonChae1104 committed:
feat(NonPreference): 마이페이지의 NonPreferenceSection 구현

- 마이페이지의 꺼려하는 음식 섹션을 구현하였습니다.

#95

HyeonChae1104 committed:
feat(FoodSelector): 음식 선택 및 확인을 위한 CategoryDropdownMenu 컴포넌트 추가

- 사용자가 카테고리별로 음식을 선택할 수 있도록 CategoryDropdownMenu 컴포넌트 구현
- 음식 이름을 표시하기 위해 `DefaultMenu` 컴포넌트 통합
- 각 음식 항목에 대해 선택 및 선택 해제 기능 추가
- 선택한 항목을 적용할 수 있는 `ConfirmButton` 추가
- `StyledFoodItem`를 통해 선택된 항목을 시각적으로 표시하도록 스타일링
- `useGetCategory`와 `useGetFoodsByCategory` 훅을 사용하여 동적 데이터 가져오기 기능 구현하였습니다.

#95

HyeonChae1104 committed:
feat(Mypage): Preference, NonPreference, FoodSelector 섹션 추가

#95

HyeonChae1104 committed:
refactor(GetFood): useGetFoodsByCategory 훅의 옵션 및 데이터 구조 수정

- `getFoodsByCategory` 함수에서 API 응답 구조 변경에 따라 `response.data.data` 대신 `response.data`로 수정
- `useGetFoodsByCategory` 훅에 `enabled` 옵션을 추가하여 선택된 카테고리가 유효할 때만 데이터 요청을 수행하도록 개선
- `options` 파라미터를 추가하여 외부에서 `enabled` 옵션을 유연하게 설정할 수 있도록 변경하였습니다.

#95

Da-Yeon-Kim commented on src/api/mock/food.mock.ts:
```diff
@@ -0,0 +1,97 @@
+    );
+  }),
+
+  rest.get(getCategoryPath(), (_req, res, ctx) => {
```
`meeting.mock.ts` 파일에 이미 카테고리 조회 관련한 목 핸들러가 있었던 걸로 기억해요! food 쪽에 두는 게 맞는 것 같아 보이니, 기존에 있던 걸 삭제하는 편이 좋아 보이네요~ 작성한 @geongyu09 씨와 상의하시길 바랍니다😃

아, 그리고 목 핸들러 데이터를 상수로 관리하신 점이나 코드가 굉장히 깔끔해서 좋네요! 다만, 다른 핸들러에서는 `status`와 `message`도 함께 상수로 관리하고 있긴 합니다ㅎㅎ 지금 코드도 충분히 예쁘지만 서로 맞추면 더 좋긴 할 것 같아요~

HyeonChae1104 replied:
감사합니다! 수정해보았습니다!!

Da-Yeon-Kim commented on src/api/hooks/useAddFoodPreference.ts:
```diff
@@ -0,0 +1,27 @@
+
+  const addFoodPreference = useMutation({
+    mutationFn: async (food: Food) => {
+      const response = await fetchWithToken.post(`/api/${type === 'preferred' ? 'preferences' : 'non-preferences'}`, {
```
백엔드와의 소통 주소는 `/api/~`로 작성하지 마시고, 만들어둔 `baseURL`을 사용해서 작성해주세요! 이 경우에는 `${baseURL}/${type ~~~}`라고 작성해야겠네요~

그리고 목 핸들러에서 이 경로를 다시 사용하기 때문에 따로 함수로 분리하는 것이 좋아 보입니다ㅎㅎ 쿼리문도 분리해주면 좋습니다~

다른 훅들은 잘 작성되었는데, 이 훅만 분리가 안 되어 있네요😊

geongyu09 replied:
동감입니다.
추가로 하나의 훅에서 여러 동작을 하는 것 같습니다.
여기서는 type을 인자로 받아서 다른 키를 주고, 다른 주소로 요청을 보내는 것 같은데, 
이 둘을 분리해보는 것은 어떨까요?

HyeonChae1104 replied:
반영하였습니다! 감사합니다

Da-Yeon-Kim commented on src/mocks/browser.ts:
```diff
@@ -1,11 +1,15 @@
 
 import { calendarMockHandler } from '@/api/mock/calendar.mock';
+import { foodMockHandler } from '@/api/mock/food.mock'
+import { handlers } from '@/api/mock/handler';
```
혹시 `handlers`는 왜 추가하신 건지 알 수 있을까요?

HyeonChae1104 replied:
원래는 추가 훅과 삭제 훅이 제대로 실행되는지 확인해보고자 추가했었는데,,
제가 이 두 훅과 함께 전에 미리 짜놓았던거라 방식도 잘못되고, 음식 관련 mock 파일이 한 개인게 맞는 것 같아서 
삭제하고 food.mock.ts에 옮겨두었습니다!

Da-Yeon-Kim commented on src/api/mock/food.mock.ts:
```diff
@@ -0,0 +1,97 @@
+
+import { baseURL } from '../instance';
+
+const getPreferencePath = () => `${baseURL}/preferences`;
```
이미 훅들에서 작성하였기 때문에 굳이 다시 새로 작성하지 말고, 핸들러에서는 import해서 사용하기만 하면 될 것 같습니다~

HyeonChae1104 replied:
감사합니다! 수정하였습니다!

Da-Yeon-Kim commented on src/components/features/MyPage/FoodSelector/index.tsx:
```diff
@@ -0,0 +1,138 @@
+
+      <FoodList>
+        {foodsByCategory?.map((food) => (
+          <StyledFoodItem
```
혹시 이미 만들어진 `defaultMenu`를 또 감싸서 스타일링하신 이유가 있나요? 클릭 시의 스타일링 때문이라면, 아예 common에 새로운 스타일링의 menu를 만드시는 게 좋아 보입니다!

Da-Yeon-Kim commented on src/components/features/MyPage/NonPreference/index.tsx:
```diff
@@ -0,0 +1,145 @@
+            {(food) => (
+              <FoodContainer key={food.food_id}>
+                <DefaultMenu menuName={food.name} />
+                <DeleteButton onClick={() => handleDeleteFood(food.food_id)}>×</DeleteButton>
```
여기서도 `DefaultMenu`와 `DeleteButton`을 감싼 컴포넌트를 새로 만드셨는데, 그보다는 common에 새로운 Menu 컴포넌트를 만드는 편이 좋아 보입니다!

Da-Yeon-Kim reviewed (commented):
제가 Join, Edit 페이지에서 해당 컴포넌트들을 재사용해야 하는데, 그렇게 하기 쉽지 않은 코드 같습니다😥 상태 업데이트나 제출 훅 같은 경우에는 그대로 이곳에서 하시고, 공통 UI, FoodSelector 같은 경우에는 공통 컴포넌트로 분리해주실 수 있나요?

또, 개인적으로 FoodSelector는 모달 팝업 같은 형식으로 보여지면 더 좋을 것 같습니다...

geongyu09 commented on src/api/mock/handler.ts:
```diff
@@ -0,0 +1,51 @@
+
+export const handlers = [
+  // 'preferences' 엔드포인트 모킹 (추가 요청 핸들러)
+  rest.post('/api/preferences', async (req, res, ctx) => {
```
이처럼 작성하게 되면 실제로 요청하는 주소와 다르게 되어 올바르게 모킹이 안되는 것 같습니다.
이전에 작성한 방식대로 수정해야합니다

geongyu09 commented on src/components/features/MyPage/FoodSelector/index.tsx:
```diff
@@ -0,0 +1,138 @@
+  const { data: categories } = useGetCategory();
+  const [selectedCategory, setSelectedCategory] = useState<string | null>(null);
+  const [selectedFoods, setSelectedFoods] = useState<Food[]>([]);
+  const { data: foodsByCategory } = useGetFoodsByCategory(selectedCategory || undefined, {
```
지금까지 제대로된 로더와 에러 컴포넌트를 만들지 않았지만, 다른 곳에서는 전부 글자로라도 적용이 되어있습니다. 
서버 상태의 상태값에 따라서 다른 UI를 보여줘야할 필요성이 있어보입니다.
현재는 !data이면 아에 아무런 UI도 보여주지 않아서 UX 적으로나, 코드 통일성 적으로나 좋지 않아보입니다 
다만, 서비스에 문제를 일으킬정도의 문제는 아니므로 나중에 시간날 때 수정해도 좋아보입니다!

geongyu09 commented on src/api/hooks/useGetNonPreferenceFoods.ts:
```diff
@@ -0,0 +1,18 @@
+  return useQuery({
+    queryKey: ['nonPreferenceFoods'],
+    queryFn: getNonPreferenceFoods,
+    staleTime: 1000 * 60 * 60 * 24, 
```
혹시 staleTime을 하루로 잡은 이유가 있나요? 
추가로 해당값이 수정될 경우가 있나요? 본인의 비선호 음식을 추가하면 다시 가져와야 하는데, 해당 부분에 대한 코드가 존재하나요?

geongyu09 commented on src/pages/MyPage/index.tsx:
```diff
@@ -1,12 +1,15 @@
 import { ProfileSection } from '@/components/features/MyPage/ProfileSection';
 
-
 export const MyPage: React.FC = () => {
```
해당 pr 내용은 아니지만, export 문이 두개 있네요
아래에 있는 export문은 지워야할 것 같습니다

`export default MyPage;`

geongyu09 commented on src/api/hooks/useAddFoodPreference.ts:
```diff
@@ -0,0 +1,27 @@
+      return response.data;
+    },
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey });
```
혹시 해당 쿼리의 캐시를 지우는 이유가 있나요? 
다른 곳에서 해당 쿼리키를 사용하지 않는 것 같습니다

HyeonChae1104 replied:
수정하였습니다!

HyeonChae1104 replied:
근데 사실은 이 부분을 통해 쿼리가 무효화 되면서 최신 데이터를 자동으로 가져오게 된다고 해서
위에서 질문하셨던 선호/ 비선호 음식을 추가하면 다시 가져오는 역할을 할 수 있을 거라고 생각을 했는데,, 

어떻게 하면 좋을까요..

HyeonChae1104 replied:
선호/ 비선호 조회 훅에 쿼리키를 두고
추가와 삭제 훅에서 쿼리를 무효화 함으로써 최신 정보를 불러오는 기능을 하고자하였는데, 
기존의 방식에는 문제가 있었던 것 같아 수정해두었습니다!

geongyu09 commented on src/api/hooks/useDeleteFoodPreference.ts:
```diff
@@ -0,0 +1,22 @@
+      return response.data;
+    },
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey });
```
여기도 마찬가지로 사용하지 않는 쿼리키를 지우고 있네요

HyeonChae1104 replied:
수정하였습니다!

geongyu09 reviewed (changes requested):
수고하셨습니다. 
시간을 많이 들이신 만큼 좋은 코드가 나왔네요

몇가지 수정이 필요해 보이는 부분이 보여서 해당 부분 남겨봤습니다

추가적으로, 저희가 기획단에서 모달이 나오는 방식으로 기획을 했었던 것 같은데, 모달 관련 로직은 없는 것 같습니다.
해당 부분은 전체 회의에서 이야기를 해봐야 할 것 같네요!

아마 실제 api로 테스트를 할 테지만, 그 전에 제대로 돌아가는지는 확인해야 하므로, msw로직 수정은 필수적으로 되어야 할 것 같습니다. 
헤당 부분 수정이 필요합니다

HyeonChae1104 committed:
chore(MyPage): 불필요한 export 문 삭제

#95

HyeonChae1104 committed:
refactor(hooks): 선호/비선호 음식 추가를 위한 훅 수정

- `type`을 사용하는 대신, `useAddPreferenceFood`와 `useAddNonPreferenceFood` 두 개의 훅으로 분리
-경로 함수로 분리
-쿼리문 분리
- 캐시 무효화가 불필요하므로 `onSuccess`에서 `invalidateQueries` 호출을 제거

#95

HyeonChae1104 committed:
refactor(hooks): 선호/비선호 음식 삭제 훅 수정

- 경로 함수 분리
- 쿼리 함수 분리
- type 사용하지 않고 두개의 훅으로 관리
- 캐시 무효화 제거

#95

HyeonChae1104 committed:
chore(hooks): export 키워드 추가 및 category 기본값 추가

- getFoodsByCategoryPath 함수에 category 기본값을 추가
- export 키워드 추가

#95

HyeonChae1104 committed:
refactor(hooks): 선호/비선호 음식 조회 훅 수정

- gcTime, staleTime 1년으로 수정

#95

HyeonChae1104 committed:
feat(FoodSelector): Modal 구현

- 모달에서 드롭다운으로 카테고리 선택
- 메뉴 선택 후 '선택완료'버튼을 통해 기존 UI에 메뉴 추가

#95

HyeonChae1104 committed:
refactor(mock): 기존의 목 데이터 형식에 맞춰 수정

- 경로 import 방식 적용
- handler.ts 삭제하고 food.mock.ts 파일에 음식 추가, 삭제 요청 핸들러 작성

#95

HyeonChae1104 committed:
feat(Menu): UI 구현을 위해 선택된 메뉴 표시, 선택 가능 메뉴 목록 컴포넌트 생성

- AddedMenu 컴포넌트 추가하여 선택된 메뉴 표시 기능 구현
- SelectableMenu 컴포넌트 추가하여 선택 가능한 메뉴 목록 구현

#95

HyeonChae1104 committed:
chore(AddedMenu): type 이름 수정

- DefaultMenu는 존재하므로 AddMenu로 수정

#95

HyeonChae1104 committed:
feat(FoodPreferenceSection): 공통 음식 선호/비선호 섹션 UI 컴포넌트 추가

- 기존의 음식 선호 및 비선호 섹션을 공통 UI 컴포넌트로 분리하여 재사용 가능하게 함
- 음식 추가 및 삭제 기능 포함, 모달 연동

#95

HyeonChae1104 committed:
refactor(FoodSelectorModal ): FoodSelector의 모달 기능을 FoodSelectorModal 컴포넌트로 분리

- 음식 선택 모달을 독립적인 `FoodSelectorModal` 컴포넌트로 분리하여 재사용성을 높임
- `FoodPreferenceSection`에서 모달 기능을 호출하여 사용 가능하게 개선
- 기존의 FoodSelector 파일 삭제

#95

HyeonChae1104 committed:
refactor(Preference, NonPreference): 선호/비선호 음식 컴포넌트를 공통 FoodPreferenceSection 컴포넌트로 리팩토링

- 선호하는 음식과 꺼려하는 음식 컴포넌트의 공통 기능을 `FoodPreferenceSection`으로 추출

#95

HyeonChae1104 commented:
리뷰에 달아주신 내용은 이렇게 반영해보았습니다! 

<구성 컴포넌트>
- `AddedMenu` : UI에 추가된 메뉴를 나타낼 때 사용
- `SelectableMenu` : 드롭다운에서 카테고리 선택 시 선택 가능한 메뉴를 나타낼 때 사용 
- `FoodPreferenceSection` : 기존의 `Preference/index.tsx`, `NonPreference/index.tsx` 파일에서 있던 공통 UI를 추출하여 재사용이 가능하도록함(이 부분에서 시간이 오래걸렸는데,, 확인 부탁드려용..)
- `FoodSelectorModal` : 모달 컴포넌트

<이외 수정 사항>
- 기존의 handler.ts 파일 삭제 후 food.mock.ts에 추가 & 기존의 mock 형식과 맞춤
- 음식 추가 훅/ 삭제 훅 type 대신 선호음식/비선호 음식 각각 다루는 두개의 훅으로 분리
- staleTime 수정

HyeonChae1104 committed:
feat(hooks): 선호 및 비선호 음식 추가, 삭제 훅에 쿼리 무효화 로직 추가하여 최신 데이터를 자동으로 가져오도록 수정

- 비선호 음식 조회 훅 쿼리키 수정
- 쿼리 무효화 로직 추가

#95

Da-Yeon-Kim commented on src/api/hooks/useAddFoodPreference.ts:
```diff
@@ -0,0 +1,45 @@
+import { useMutation, useQueryClient } from '@tanstack/react-query';
```
Preferences와 nonPreferences는 다른 파일로 분리하시면 좋을 것 같습니다! 동작이 유사하긴 하지만, 독립적이니까요😀

Da-Yeon-Kim commented on src/api/hooks/useDeleteFoodPreference.ts:
```diff
@@ -0,0 +1,42 @@
+import { useMutation, useQueryClient } from '@tanstack/react-query';
```
여기도 마찬가지로 따로 파일을 두는 게 좋은 것 같습니다!

Da-Yeon-Kim commented on src/components/common/Food/FoodPreferenceSection/index.tsx:
```diff
@@ -0,0 +1,111 @@
+  border-bottom: 1px solid ${colors.gray};
+`;
+
+const Spacing = styled.div<{ height: number }>`
```
여기 spacing은 이미 만들어진 게 있으니 활용하시면 될 것 같아요! `common/layouts/Spacing`에 있습니다~

Da-Yeon-Kim commented on src/components/common/Food/FoodPreferenceSection/index.tsx:
```diff
@@ -0,0 +1,111 @@
+  refetchFoods: () => void;
+};
+
+export const FoodPreferenceSection: React.FC<Props> = ({ title, foods, onAddFood, onDeleteFood, refetchFoods }) => {
```
음, 이렇게 자주 반복되는 컴포넌트 묶음을 common으로 만드신 건 좋은 시도 같습니다! 동작이 원활하게 된다면, 제가 유용하게 사용할 수 있겠네요👍✨

그러나 지금 API 요청을 통해 받아온 데이터 foods를 `selectedFoods` 상태와 동기화해 화면상에 기본값으로 보여주는 로직이 없네요!

![image](https://github.com/user-attachments/assets/43bba32e-5b2e-4d65-9db2-0f2b41d96e49)
(예시: 불고기와 스테이크를 API로 받아왔으나 아무것도 없는 기본 화면....)

HyeonChae1104 replied:
해결했습니다!

Da-Yeon-Kim commented on src/components/common/Food/FoodSelectorModal/index.tsx:
```diff
@@ -0,0 +1,147 @@
+export const FoodSelectorModal: React.FC<Props> = ({ onFoodSelect, onClose }) => {
+  const { data: categories } = useGetCategory();
+  const [selectedCategory, setSelectedCategory] = useState<string | null>(null);
+  const [selectedFoods, setSelectedFoods] = useState<Food[]>([]);
```
`FoodSelectorModal`에서 사용하는 `selectedFoods`는 `FoodPrefrenceSection`의 `selectedFoods`와 독립적이네요.

그래서 지금 모달 창을 닫았다가 다시 열면, 선택한 메뉴가 선택했다고 표시되지 않는 것 같습니다.

![image](https://github.com/user-attachments/assets/163dad14-1c48-4b19-b5c1-43bd80a8f179)
![image](https://github.com/user-attachments/assets/4c86615b-1ef5-45b3-9e33-49d19be23ee8)

(예시: 이미 선택한 한식이 모달 창에서는 선택 X로 표시. 그래서 추가하면 두 개가 되는 것이 보입니다)

일단 제일 먼저 생각난 해결책은 `useForm` 이용해서 전역 상태로 관리하는 것인데,, 지금 이 컴포넌트들이(`FoodPrefrenceSection`, `FoodSelectorModal`) common으로 여러 곳에서 사용되기 때문에 자칫 잘못하면 모든 곳에서 상태가 공유될지 모른다는 불상사가.... `useForm`에서 독립적으로 관리하도록 방지할 수는 있을 것 같긴 합니다....

하지만 영 그렇다면 props로 내려주는 방법도 있긴 한 것 같아요....

HyeonChae1104 replied:
반영하였습니다! 감사합니다

Da-Yeon-Kim reviewed (commented):
이 부분 구현 때문에 고생이 많으시네요!
점점 더 좋은 코드가 되어가는 것 같습니다👍👍 쑥쑥 깔끔하고 예뻐지는 코드~~

자기 전에 동작이 잘 되는지만 빠르게 훑어봤는데, 몇 가지 안 되는 게 있는 것 같아 보이네요! 일단 이 부분들을 수정하셔야 할 것 같습니다~~😀

사실 저도 어떤 게 일반적인 코드인지 잘 몰라서 모든 팀원을 대상으로 한 가지 물어보고 싶은데요! `FoodPreferenceSection`을 아예 common하게 만드는 편이 좋은가요? 많은 페이지에서 사용되긴 하지만, 많은 동작을 가지고 있어서 아예 한 묶음으로 만드는 건 조금 어려워 보입니다.... 그리고 UI가 아닌 동작을 포함해도 common에 있는 게 맞는지도 궁금하고요!

(일단 저는 자러 갑니다...🛏 다들 좋은 꿈 꾸세요~~)

HyeonChae1104 committed:
refactor(hooks): 선호/비선호 음식 추가 및 삭제 훅 파일 분리

- AddNonPreference, AddPreference, DeleteNonPreference, DeletePreference로 파일을 분리하였습니다.

#95

HyeonChae1104 committed:
refactor(hooks): 마이페이지에서 선호/비선호 조회 값이 나타나지 않던 문제 해결

- fix(hooks): 비선호 음식 데이터가 마이페이지에 표시되지 않던 문제 해결

- 선호/비선호음식 조회 함수에서 잘못된 데이터 경로 (`response.data.data`)를 `response.data`로 수정
- API 응답 구조에 맞게 데이터 반환하여 UI에 정상적으로 값 표시되도록 수정

#95

HyeonChae1104 committed:
refactor(FoodPreferenceSection): Spacing import 해오는 것으로 수정

#95

HyeonChae1104 committed:
refactor(): selectedFoods가 독립적이었던 문제를 props로 해결

- FoodPrefrenceSection, FoodSelectorModal에서 selectedFoods가 독립적이었던 문제를 props로 해결

#95

HyeonChae1104 committed:
feat(FoodSelectorModal): 모달에 검색창 추가

- 모달의 드롭다운 하단에 검색창을 추가하였습니다.
- 검색기능은 카테고리를 선택하면 그 카테고리에 속하는 메뉴가 검색 가능합니다.

#95

geongyu09 commented on src/api/hooks/useGetPreferenceFoods.ts:
```diff
@@ -0,0 +1,19 @@
+  return useQuery({
+    queryKey: ['preferenceFoods'],
+    queryFn: getPreferenceFoods,
+    gcTime: 1000 * 60 * 60 * 24 * 365, 
```
와우.. 이렇게 길게 작성한 이유가 있을까요?

geongyu09 commented on src/api/hooks/useAddNonPreferenceFood.ts:
```diff
@@ -0,0 +1,26 @@
+
+export const getAddNonPreferenceFoodPath = () => `${baseURL}/non-preferences`;
+
+
```
여기 prettier가 적용이 안되어 있는 것 같습니다

geongyu09 commented on src/api/hooks/useAddPreferenceFood.ts:
```diff
@@ -0,0 +1,25 @@
+  return useMutation({
+    mutationFn: addPreferenceFood,
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ['preferenceFoods'] });
```
좋네요 
시간적으로 여유가 된다면 낙관적 업데이트를 적용해봐도 좋아 보이지만, 지금도 충분히 좋은 코드입니다

geongyu09 commented on src/api/hooks/useGetFood.ts:
```diff
@@ -0,0 +1,24 @@
+export const getFoodsByCategoryPath = (category: string = '') => `${baseURL}/foods?category=${category}`;
+
+const getFoodsByCategory = async (category: string): Promise<Food[]> => {
+  if (!category || category === 'undefined') {
```
카테고리가 undefined가 되는 경우는 없어보이네요. 
단순히 빈 배열만 받는 경우만 있을 것 같아요

geongyu09 commented on src/api/hooks/useGetFood.ts:
```diff
@@ -0,0 +1,24 @@
+
+export const getFoodsByCategoryPath = (category: string = '') => `${baseURL}/foods?category=${category}`;
+
+const getFoodsByCategory = async (category: string): Promise<Food[]> => {
```
반환 타입을 지정하지 않아도 올바르게 추론될 것 같습니다

물론 이를 위해서는 기존의 코드들처럼 .get<type>(); 이런식으로 작성해야합니다

꼭 수정해아 하는 부분은 아니라고 생각합니다

geongyu09 commented on src/api/hooks/useGetFood.ts:
```diff
@@ -0,0 +1,24 @@
+  return useQuery<Food[]>({
+    queryKey: ['foods', category],
+    queryFn: () => getFoodsByCategory(category!),
+    enabled: options?.enabled ?? !!category,
```
enable은 결국에 categroy값이 있을때만 요청하도록 되어있는데, 더 추가적으로 enable조건이 필요한가요?

geongyu09 commented on src/components/common/Food/FoodSelectorModal/index.tsx:
```diff
@@ -0,0 +1,170 @@
+  const { data: categories } = useGetCategory();
+  const [selectedCategory, setSelectedCategory] = useState<string | null>(null);
+  const { data: foodsByCategory } = useGetFoodsByCategory(selectedCategory || undefined, {
+    enabled: !!selectedCategory,
```
카테고리값이 존재할 때만 도요청하도록 되어있네요
해당 부분은 조금 더 간소화 할 수 있을 것 같은데, 지금 잘 동작하니 괜찮아보입니다

geongyu09 commented on src/components/features/MyPage/NonPreference/index.tsx:
```diff
@@ -0,0 +1,33 @@
+  const addNonPreferenceFood = useAddNonPreferenceFood();
+  const deleteNonPreferenceFood = useDeleteNonPreferenceFood();
+
+  const handleAddFood = (food: Food) => addNonPreferenceFood.mutate(food);
```
addNonPreferenceFood.가 반복되는 것 같습니다

Da-Yeon-Kim reviewed (commented):
음, API 소통을 선호/비선호 음식으로 분리하니까 좋네요~
지금 잠깐 동작만 먼저 확인해봤는데, 이번에도 2가지가 구현이 안 되고 있습니다!

1. 모달 창에서 선택 취소가 안 됩니다.
2. 모달 창을 통해 선택한 건 모달 창에서 선택 표시가 되어 있는데, API를 통해 가져와서 기본값으로 주어지는 건 선택되었다는 표시가 안 뜨네요. 삭제할 때, 기본값 불고기와 모달로 선택한 불고기가 따로 삭제되는 걸로 봐서 아마 아예 따로 관리되는 것 같아요. 저번부터 상태 관리에서 문제가 생기는 것 같네요😥

구현하시면서, 여러 가지 케이스로 동작해보면서 기능이 잘 되고 있는지 점검하시면 더 좋을 것 같아요~

+) 추가로, 지금 검색창이 카테고리 선택 전에도 뜨고 있는데 이러면 사용자 입장에서는 전체 검색도 되는 줄 알고 헷갈릴 수 있을 것 같아요. 검색창은 카테고리가 선택되고 나면 뜨게 하는 편이 좋을 것 같습니다. (아니면 카테고리 검색 전에는 disable로 설정해도 괜찮겠네요)

geongyu09 commented on src/components/features/MyPage/Preference/index.tsx:
```diff
@@ -0,0 +1,33 @@
+  const handleDeleteFood = (foodId: number) => deleteFoodPreference.mutate(foodId);
+
+  useEffect(() => {
+    refetch();
```
혹시 여기서 사이드이팩트로 첫 마운트때 한번 더 요청하는 이유가 있나요?

geongyu09 reviewed (approved):
코드 양이 정말 많네요..
일단 동작하므로 approve합니다
남긴 커맨트들은 참고만해주세요

HyeonChae1104 committed:
fix: 코드 리뷰 반영

- 모달에서 선택 취소 가능
- mock데이터 수정 및 코드 수정으로 foodID를 기준으로 음식 조회, 추가, 삭제가 이루어지도록함.
- 검색창이 카테고리 선택 후 나오도록 변경

#95

HyeonChae1104 commented:
아까부터 2번으로 적어주신 내용을 계속 해보고 있었는데 안되어서 일단 리뷰를 받았었는데
설마 하면서 목데이터를 보니,, 조회때의 음식ID와 추가때의 음식ID가 다른걸로 적혀있더라구요..
그 부분 수정과 함께 음식ID를 기준으로 상태관리하는 것으로 해두었습니다.

추가로 모달에서 선택, 선택해제 가능하게 하였고, 검색창도 카테고리 선택 후 보여지도록 변경하였습니다! 디테일한 부분까지 말해주셔서 감사합니다.!

Da-Yeon-Kim reviewed (approved):
와, 이제 잘 동작하네요! 와아~~~😃
동작이 잘 되니까 approve 하고 저도 사용해서 구현하도록 하겠습니다!
사용하면서 어려운 점 있으면 질문할게요!
